File: php/tests/SupermarketTest.php
<?php

declare(strict_types=1);

namespace Tests;

use PHPUnit\Framework\TestCase;
use Supermarket\Model\Product;
use Supermarket\Model\ProductUnit;
use Supermarket\Model\ShoppingCart;
use Supermarket\Model\SpecialOfferType;
use Supermarket\Model\Teller;

class SupermarketTest extends TestCase
{
    public function testTenPercentDiscount(): void
    {
        // Arrange
        $catalog = new FakeCatalog();
        $toothbrush = new Product('toothbrush', ProductUnit::EACH());
        $catalog->addProduct($toothbrush, 0.99);
        $apples = new Product('apples', ProductUnit::KILO());
        $catalog->addProduct($apples, 1.99);

        $cart = new ShoppingCart();
        $cart->addItemQuantity($apples, 2.5);

        $teller = new Teller($catalog);
        $teller->addSpecialOffer(SpecialOfferType::TEN_PERCENT_DISCOUNT(), $toothbrush, 10.0);

        // Act
        $receipt = $teller->checkoutArticlesFrom($cart);

        // Assert
        self::assertSame(4.975, $receipt->getTotalPrice());
        self::assertSame([], $receipt->getDiscounts());
        self::assertCount(1, $receipt->getItems());
        $receiptItem = $receipt->getItems()[0];
        self::assertSame($apples, $receiptItem->getProduct());
        self::assertSame(1.99, $receiptItem->getPrice());
        self::assertSame(2.5 * 1.99, $receiptItem->getTotalPrice());
        self::assertSame(2.5, $receiptItem->getQuantity());
    }
}

File: dart/supermarket_receipt/test/supermarket/model/supermarket_test.dart
import 'package:supermarket_receipt/supermarket/model/product.dart';
import 'package:supermarket_receipt/supermarket/model/product_unit.dart';
import 'package:supermarket_receipt/supermarket/model/receipt.dart';
import 'package:supermarket_receipt/supermarket/model/receipt_item.dart';
import 'package:supermarket_receipt/supermarket/model/shopping_cart.dart';
import 'package:supermarket_receipt/supermarket/model/special_offer_type.dart';
import 'package:supermarket_receipt/supermarket/model/supermarket_catalog.dart';
import 'package:supermarket_receipt/supermarket/model/teller.dart';
import 'package:test/test.dart';

import '../fake_katalog.dart';

void main() {
  group('SupermarketTest', () {
    test('tenPercentDiscount', () {
      SupermarketCatalog catalog = FakeCatalog();
      Product toothbrush = Product('toothbrush', ProductUnit.EACH);
      catalog.addProduct(toothbrush, 0.99);
      Product apples = Product('apples', ProductUnit.KILO);
      catalog.addProduct(apples, 1.99);

      Teller teller = Teller(catalog);
      teller.addSpecialOffer(SpecialOfferType.TEN_PERCENT_DISCOUNT, toothbrush, 10.0);

      ShoppingCart cart = ShoppingCart();
      cart.addItemQuantity(apples, 2.5);

      Receipt receipt = teller.checksOutArticlesFrom(cart);

      expect(receipt.getTotalPrice(), closeTo(4.975, 0.01));
      expect(receipt.getDiscounts(), isEmpty);
      expect(receipt.getItems().length, 1);
      ReceiptItem receiptItem = receipt.getItems()[0];
      expect(receiptItem.getProduct(), apples);
      expect(receiptItem.getPrice(), 1.99);
      expect(receiptItem.getTotalPrice(), 2.5 * 1.99);
      expect(receiptItem.getQuantity(), 2.5);
    });
  });
}